storage/innobase/buf/buf0flu.cc

```diff
@@ -2394,7 +2428,7 @@ static void buf_flush_page_cleaner()
       if (oldest_lsn >= soft_lsn_limit)
         buf_flush_async_lsn= soft_lsn_limit= 0;
     }
-    else if (buf_pool.ran_out())
+    else if (buf_pool.need_LRU_eviction())
```

Comment:
Not related to this patch but just thought of mentioning as it could affect. 

The code below has unconditional loop over flush list. This could be expensive when the buffer pool is large and expected number of dirty pages in the system is very large. Since this patch would lead to more number of LRU flush it could start impacting.

Reply:
The code I am referring to. Looks like some clean-up by unconditionally looping over flush list. 
```c
/* Remove clean blocks from buf_pool.flush_list before the LRU scan. */
      for (buf_page_t *p= UT_LIST_GET_FIRST(buf_pool.flush_list); p; )
      {
        const lsn_t lsn{p->oldest_modification()};
        ut_ad(lsn > 2 || lsn == 1);
        buf_page_t *n= UT_LIST_GET_NEXT(list, p);
        if (lsn <= 1)
          buf_pool.delete_from_flush_list(p);
        p= n;
      }
```

Reply:
That scan was added in 9d1466522ea92963ac6ca16b597392714280c9f1 in order to prevent a race condition at the end of crash recovery. This is also related to 5b53342a6a59ae1141b2a46467954fc3836f8e80. Ever since 22b62edaedddb1cabd5b855cdd39a5e90a5695a2 we do not remove blocks from `buf_pool.flush_list` on `buf_page_write_complete()`, but only update the `oldest_modification` to the special value 1 that denotes that the page is actually clean. In this way, we significantly reduced contention on the `buf_pool.flush_list_mutex`.

Reply:
ok, thanks for explaining. I will go through the optimization sometime.

Reply:
The loop that you identified looks redundant and does now cause a significant regression. #3078 was filed for the removal.